### ISSUE TITLE
Justification (OT JSTF/AAT just/fallback)

### ISSUE THREAD

**Issue by ebraminio** [harfbuzz/harfbuzz#1469]
Now that Edge wants to switch to Chromium, they will lose their Kashida justification (cross filed in crbug.com/913847 ) so HarfBuzz and Chromium should prioritize this non-trivial issue to fix, which is in 2019 roadmap HarfBuzz.

Perhaps Microsoft, now a Chromium user, can share their experience on the matter also.

(cc @drott @kojii @eaenet @petercon)

**Comment by ebraminio** [comment]
See #586 also for more context

**Comment by khaledhosny** [comment]
I think the way Edge/IE/MS Word do it is basically similar to what LibreOffice does (or rather LibreOffice emulates MS implementation) which is done entirely at character level by analyzing the text and inserting U+0640 characters as needed without reshaping. This approach has serious limitation of course, but if the intent is to preserve what Edge/IE are doing then it should be enough. 

OpenType JUST table is yet to be proven as a usable table as there is no fonts or layout engines that support it, so even if it was implemented today it will not help in any meaningful way. The AAT just table is probably in a better position since some Apple API actually use it, but that will be limited to the handful Arabic AAT fonts that have it, so still of not much value in practice.

So I’d say Chrome should first do the character level implementation then figure out what kind of API they would need, instead of implementing an API blindly that potentially no one will be able to use.

**Comment by ebraminio** [comment]
Right. Arial and Times New Roman are the only JSTF having fonts shipped with Windows and macOS and the table. `just`, as noted, is shipped with Courier and most of Arabic fonts in macOS, but also LucidaGrande (no Arabic AFAIK) https://gist.github.com/ebraminio/d432e831b3f7ebe30245dde5775e1c7e

Making Chrome to have that first seems like a good plan I believe also.

The way and place DirectWrite can apply justification https://github.com/harfbuzz/harfbuzz/blob/master/src/hb-directwrite.cc#L702-L778 can be inspiration for Chromium or HarfBuzz I think.

**Comment by ebraminio** [comment]
And worth to note I guess, Internet Explorer had an interesting bit, "text-kashida-space" https://technet.microsoft.com/en-us/ms531173(v=vs.71) later deprecated but still working, which could control kashida/space percentage for justification, maybe something worthy for a higher-level text rendering library such as Pango.

**Comment by ebraminio** [comment]
Am working on something [here](https://github.com/ebraminio/harfbuzz/tree/kashida), interestingly that whole hb-ot-shape-complex-arabic won't be ran on AAT fonts (well, obviously) my main question is, can we rely to `just` table presence on AAT fonts, or, we should still do Kashida justification when a font is morx but doesn't have `just`? Considering how both AAT and the need Kashida are little, a possible answer can be "nobody cares", I thought however thinking about that can help to better understanding of the API or the work.

**Comment by simoncozens** [comment]
I'm starting to look at JSTF support in SILE, but I'm confused about what level in the layout stack this occupies. The OT spec tells you how to represent the justification data, but doesn't give you much clue about how to use it.

In the absence of Harfbuzz support I'm planning to have SILE consume the JSTF table itself, but it has got me thinking about how we expect JSTF support to work in the future. Should Harfbuzz report possible expansion/contractions and substitutions to the layout engine so that the layout engine can do the substitutions itself, or should the layout engine give Harfbuzz information about line shortfalls, so that Harfbuzz can just return a stream of glyph IDs as normal?

**Comment by behdad** [comment]
Not what this issue is asking for, but now there's experimental `hb_shape_justify` that uses a variable-font axis to perform justification.
